### PR TITLE
Issue 983 half cell

### LINKED ISSUE #983: Half-cell model for Li-ion battery
**Summary**
Could a half cell model (vs Li foil) be added along with the already-existed full cell model for Li-ion battery? It could be useful to simulate lab tests as they're frequently built in this configuration.

### PR DESCRIPTION
**Pull request by ferranbrosa**
# Description

Basic DFN model for a half-cell. This is an experimental functionality which doesn't support all PyBaMM features (e.g. experiments) and which will be integrated into the submodel structure in the future.

Fixes #983 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/expression_tree/binary_operators.py [hunk 1]
```diff
@@ -730,6 +730,21 @@ def _binary_evaluate(self, left, right):
...
+
+    def __str__(self):
+        """ See :meth:`pybamm.Symbol.__str__()`. """
+        return "{!s} mod {!s}".format(self.left, self.right)
+
+    def _binary_evaluate(self, left, right):
+        """ See :meth:`pybamm.BinaryOperator._binary_evaluate()`. """
+        return left % right
+
+
```
Comment (tinosulzer) [inline pybamm/expression_tree/binary_operators.py:747]: this needs a `jac` and a `diff` (well, I guess casadi takes care of it, but may as well add now)

File: pybamm/models/full_battery_models/base_battery_model.py [hunk 2]
```diff
@@ -189,6 +189,7 @@ def options(self, extra_options):
             "external submodels": [],
             "sei": None,
             "sei porosity change": False,
+            "working electrode": "negative"
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/base_battery_model.py:192]: shouldn't this be `None` by default?

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+
+
+class BasicDFNHalfCell(BaseModel):
+    """Doyle-Fuller-Newman (DFN) model of a lithium-ion battery with lithium counter
+    electrode, from [2]_.
+
+    This class differs from the :class:`pybamm.lithium_ion.DFN` model class in that it
+    shows the whole model in a single class. This comes at the cost of flexibility in
+    comparing different physical effects, and in general the main DFN class should be
+    used instead.
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:15]: rephrase this to make it clear that it's half-cell

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+           derivation of a single particle model with electrolyte”. Journal of The
+           Electrochemical Society, 166(15):A3693–A3706, 2019
+
+    **Extends:** :class:`pybamm.lithium_ion.BaseModel`
+    """
+
+    def __init__(
+        self,
+        name="Doyle-Fuller-Newman half cell model",
+        options={"working electrode": "negative"},
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:34]: I think there shouldn't be a default for this option, it should always be specified (unless negative is much more common than positive?)

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+        ######################
+        # Particles
+        ######################
+
+        # The div and grad operators will be converted to the appropriate matrix
+        # multiplication at the discretisation stage
+        N_s_n = -param.D_n(c_s_n, T) * pybamm.grad(c_s_n)
+        N_s_p = -param.D_p(c_s_p, T) * pybamm.grad(c_s_p)
+        self.rhs[c_s_n] = -(1 / param.C_n) * pybamm.div(N_s_n)
+        self.rhs[c_s_p] = -(1 / param.C_p) * pybamm.div(N_s_p)
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:198]: only need to do one of these depending on what is the working electrode

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+        # dphie_dx = (
+        #     -i_cell / (param.kappa_e(c_e, T) * tor * param.gamma_e / param.C_e)
+        #     + param.chi(c_e) * dce_dx / c_e
+        # )
+
+        if working_electrode == "negative":
+            self.boundary_conditions[phi_e] = {
+                "left": (pybamm.Scalar(0), "Neumann"),
+                "right": (pybamm.Scalar(0), "Dirichlet"),
+                # "right": (pybamm.boundary_value(dphie_dx, "right"), "Neumann"),
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:334]: remove comment

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+        # Current in the electrolyte
+        ######################
+        i_e = (param.kappa_e(c_e, T) * tor * param.gamma_e / param.C_e) * (
+            param.chi(c_e) * pybamm.grad(c_e) / c_e - pybamm.grad(phi_e)
+        )
+        self.algebraic[phi_e] = pybamm.div(i_e) - j
+        # dphie_dx = (
+        #     -i_cell / (param.kappa_e(c_e, T) * tor * param.gamma_e / param.C_e)
+        #     + param.chi(c_e) * dce_dx / c_e
+        # )
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:328]: remove comment

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+            self.boundary_conditions[phi_e] = {
+                "left": (pybamm.Scalar(0), "Neumann"),
+                "right": (pybamm.Scalar(0), "Dirichlet"),
+                # "right": (pybamm.boundary_value(dphie_dx, "right"), "Neumann"),
+            }
+        else:
+            self.boundary_conditions[phi_e] = {
+                "left": (pybamm.Scalar(0), "Dirichlet"),
+                "right": (pybamm.Scalar(0), "Neumann"),
+                # "right": (pybamm.boundary_value(dphie_dx, "right"), "Neumann"),
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:340]: remove comment

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py (hunk 6 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:334]: Done

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py (hunk 7 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:328]: Done

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py (hunk 8 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:340]: Done

File: pybamm/models/full_battery_models/base_battery_model.py (hunk 2 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/base_battery_model.py:192]: Done

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py (hunk 4 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:34]: There is no default now

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py (hunk 3 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:15]: Done

File: pybamm/expression_tree/binary_operators.py (hunk 1 above)
Comment (ferranbrosa) [reply pybamm/expression_tree/binary_operators.py:747]: Done

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py (hunk 5 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:198]: Done

File: pybamm/solvers/base_solver.py [hunk 9]
```diff
@@ -315,6 +317,32 @@ def report(string):
...
+                    elif symbol.left.id == (pybamm.t * model.timescale).id:
+                        expr = symbol.right.new_copy() / symbol.left.right.new_copy()
+                        found_t = True
+
+                    # Update the events if the modulo function depended on t
+                    if found_t:
+                        if t_eval is None:
+                            N_events = 200
+                        else:
+                            N_events = t_eval[-1] // expr.new_copy().value
```
Comment (tinosulzer) [inline pybamm/solvers/base_solver.py:336]: I don't think you need a `new_copy` here (if just taking the value)

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+        Discharge of the Lithium/Polymer/Insertion Cell”. Journal of The
+        Electrochemical Society, 140(6):1526-1533, 1993
+
+    **Extends:** :class:`pybamm.lithium_ion.BaseModel`
+    """
+
+    def __init__(
+        self,
+        name="Doyle-Fuller-Newman half cell model",
+        options={"working electrode": None},
```
Comment (tinosulzer) [inline pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py:38]: Safer to do `options=None` and then in the first line of the function
```
options = options or {"working electrode": None}
```
which is the same as
```
if options is None:
     options = {"working electrode": None}
```

File: pybamm/solvers/base_solver.py (hunk 9 above)
Comment (ferranbrosa) [reply pybamm/solvers/base_solver.py:336]: Done

File: pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py (hunk 10 above)
Comment (ferranbrosa) [reply pybamm/models/full_battery_models/lithium_ion/basic_dfn_half_cell.py]: Done
